### PR TITLE
Fix useless Dialyzer warnings by moving result spec code to separate private function

### LINKED ISSUE #85: ElixirLS Dialyzer (VS code) warning : The pattern 'false' can never match the type  'true'
Hi,  first I want to thank you for this great library , I have some issue with Dialyzer warning but I'm not sure why , here's my settings:

Mix.exs: 
```elixir
      {:type_check, github: "Qqwy/elixir-type_check"}
```

Elixir vers:
   1.12.2-otp-24

ElixirLS vs code: 
  0.8.1

Code:

```elixir
defmodule Document.ExcelTest do
  use TypeCheck
  alias Document.ExcelTest

  defstruct [
    :name,
    :age,
    :phone
  ]

  @type! t() :: %ExcelTest{
           name: String.t(),
           age: integer(),
           phone: String.t()
         }
end
```
```elixir
defmodule Document.Doc do

  use TypeCheck
  alias Document.ExcelTest

  @spec! parse_list(any) :: ExcelTest.t()
  def parse_list(list) do

    column = [
      :name,
      :age,
      :phone
    ]

    Enum.zip(column, list)
    |> Enum.into(%{})
    |> then( fn map -> struct(%ExcelTest{}, map) end)

  end
end
```
 the code can be compiled and run  , but there' warning from dialyzer ;
 
screenshot

Is there something wrong with my code ?

### PR DESCRIPTION
**Pull request by Qqwy**
Fixes #85

### CODE REVIEW

**Comment by Qqwy** [comment]
@tomekowal @relax2code Could you test whether the issues you were encountering are fixed by this PR?

(You can use this version of the library by temporarily changing the line in your mix deps to `{:type_check, github: "qqwy/elixir-type_check", branch: "85-dialyzer2"},`)

**Comment by tomekowal** [comment]
Yes, it solved that issue.

**Comment by mindreframer** [comment]
Confirming also here the issue is solved. Thanks for the quick turnaround!

File: lib/type_check/spec.ex [hunk 1]
```diff
@@ -149,6 +148,15 @@ defmodule TypeCheck.Spec do
       defoverridable([{unquote(name), unquote(arity)}])
 
       Kernel.unquote(function_kind)(unquote(name)(unquote_splicing(clean_params)), do: unquote(body))
+
+      # The result is checed in a separate function
```
Comment (mindreframer) [inline lib/type_check/spec.ex:152]: a small typo: `the result is checked`

**Comment by Qqwy** [comment]
Wonderful, thank you for the feedback!

I'll release this fix in a new version to Hex.PM ASAP :rocket: !

**Comment by mindreframer** [comment]
Also thanks for this superb library and congrats for the upcoming talk about TypeCheck at the next ElixirConf! I'm looking forward to watch as soon as possible. Cheers!

**Comment by mindreframer** [comment]
@Qqwy I found a case, where the bug seems to persist... 

```elixir
### this module works fine, because of `any` in spec: `@spec! from_db(any) :: t()`
defmodule Types.Ad do
  use TypeCheck
  defstruct id: nil, consumer_price: nil

  @type! t :: %Types.Ad{
           id: number(),
           consumer_price: number()
         }

  @spec! from_db(any) :: t()
  def from_db(row) do
    %Types.Ad{
      id: Map.get(row, "id")
    }
  end
end

### this fails, because of `any` in spec: `@spec! from_db(any) :: t()`
defmodule Types.Ad3 do
  use TypeCheck
  defstruct id: nil

  @type! t :: %Types.Ad3{
           id: number()
         }

  @spec! from_db(map) :: t()
  def from_db(row) do
    %Types.Ad3{}
  end
end
```


ERROR message:
```
lib/type_check/spec.ex:22:call_without_opaque
Function call without opaqueness type mismatch.

Call does not have expected term of type
  {{TypeCheck.Builtin.Map.t()
    | %{
        :__struct__ => atom(),
        :choices => [any()],
        :element_type => _,
        :element_types => [any()],
        :keypairs => [any()],
        :local => boolean(),
        :name => atom(),
        :range => map(),
        :type => _,
        :value => _
      }, atom(),
    %{
      :expected_length => non_neg_integer(),
      :expected_size => integer(),
      :index => integer(),
      :key => _,
      :keys => [any()],
      :problem => _,
      :problems => [any()]
    }, _}, nil | maybe_improper_list() | map()}
  | {TypeCheck.Builtin.Map.t()
     | %{
         :__struct__ => atom(),
         :choices => [any()],
         :element_type => _,
         :element_types => [any()],
         :keypairs => [any()],
         :local => boolean(),
         :name => atom(),
         :range => map(),
         :type => _,
         :value => _
       }, atom(),
     %{
       :expected_length => non_neg_integer(),
       :expected_size => integer(),
       :index => integer(),
       :key => _,
       :keys => [any()],
       :problem => _,
       :problems => [any()]
     }, _}
 (with opaque subterms) in the 1st position.

TypeCheck.TypeError.exception(
  {{%TypeCheck.Spec{
      :location => {<<_::464>>, 30},
      :name => :from_db,
      :param_types => [
        %TypeCheck.Builtin.Map{:key_type => map(), :value_type => map()},
        ...
      ],
      :return_type => %TypeCheck.Builtin.NamedType{
        :local => false,
        :name => <<_::88>>,
        :type => %TypeCheck.Builtin.FixedMap{:keypairs => [any(), ...]}
      }
    }, :param_error,
    %{
      :index => 0,
      :problem =>
        {%TypeCheck.Builtin.Map{
           :key_type => %TypeCheck.Builtin.Any{},
           :value_type => %TypeCheck.Builtin.Any{}
         }, :not_a_map, %{}, _}
    }, [any(), ...]}, [{:file, <<_::464>>} | {:line, 22}, ...]}
)

________________________________________________________________________________
done (warnings were emitted)
Halting VM with exit status 2
```

I guess there is still some fishiness going on. Hope this helps a bit to narrow down this issue.

**Comment by Qqwy** [comment]
@mindreframer I am pretty sure that warning from Dialyzer is caused by a different underlying problem.
I'll move your example to a new issue to tackle separately!

**Pull request merged**